physionet-django/project/models.py

```diff
@@ -1533,28 +1533,179 @@ def get_storage_info(self, force_calculate=True):
+            Harvard : Harvard
+            Vancouver : Vancouver
+
+        Returns
+        -------
+        citation_dict (dictionary) : dictionary of each citation style and
+            their respective citation for the requested project
+        """
+        # For if there are no authors or valuable metadata
         if self.is_legacy:
-            return ''
+            return {'APA':''}
 
+        styles = ['APA','MLA','Chicago','Harvard','Vancouver']
         authors = self.authors.all().order_by('display_order')
-        if self.doi:
-            return format_html(
-                '{authors} ({year}). {title} (version {version}). '
-                '<i>PhysioNet</i>. '
-                '<a href="https://doi.org/{doi}">https://doi.org/{doi}</a>',
-                authors=', '.join(a.initialed_name() for a in authors),
-                year=self.publish_datetime.year,
-                title=self.title,
-                version=self.version,
-                doi=self.doi)
-        else:
-            return format_html(
-                '{authors} ({year}). {title} (version {version}). '
-                '<i>PhysioNet</i>.',
-                authors=', '.join(a.initialed_name() for a in authors),
-                year=self.publish_datetime.year,
-                title=self.title,
-                version=self.version)
+        citation_dict = {}
+
+        for style in styles:
+
+            if (style == 'MLA'):
+
+                if (len(authors) == 1):
+                    temp_authors = authors[0].get_full_name(reverse=True)
+                elif (len(authors) == 2):
+                    first_author = authors[0].get_full_name(reverse=True)
+                    second_author = authors[1].get_full_name()
+                    temp_authors = first_author + ', and ' + second_author
+                else:
+                    temp_authors = ', '.join(authors[0].get_full_name().split())
+                    temp_authors += ', et al'
+
+                if self.doi:
+                    citation_format = format_html(
+                        '{authors}. "{title} (version {version})." '
+                        '<i>PhysioNet</i> ({year}), doi:'
+                        '<a href="https://doi.org/{doi}">https://doi.org/{doi}</a>.',
+                        authors=temp_authors,
+                        title=self.title,
+                        version=self.version,
+                        year=self.publish_datetime.year,
+                        doi=self.doi)
+                else:
+                    citation_format = format_html(
+                        '{authors}. "{title} (version {version})." '
+                        '<i>PhysioNet</i> ({year}).',
+                        authors=temp_authors,
+                        title=self.title,
+                        version=self.version,
+                        year=self.publish_datetime.year)
+
+            elif (style == 'APA'):
+
+                if (len(authors) == 1):
```

Comment:
is there a way to avoid the if clauses here? e.g. could the authors variable be populated with a list comprehension?

Reply:
Hmmm, could you explain further? Each citation style has a particular formatting guideline.

Reply:
I see potential for further refactoring but it might make it harder to read.

Reply:
Most information seems consistent across styles, except perhaps the author string. Perhaps we could separate the structure from the content doing something like:

```
# get the structure
apa = "{author} {year} {title} {blah}"
harvard = "{year} {author} {blah}"

# get information that is fixed for everyone
content = {"year": x,
                    "title": title}
}

# get style specific strings
if harvard: authors = x
elif apa authors = x
etc.

# unpack content with **
citation = apa.format(authors, **content)
```

Benjamin or Felipe may have a cleaner solution...